### ISSUE TITLE
Recovering formerly boundary faces in periodic meshes

### ISSUE THREAD

**Issue by logantm2** [mfem/mfem#3193]
We're looking into adding periodic boundaries to our particle-based app. It looks to be pretty straightforward to "periodic-ize" meshes using the [CreatePeriodicVertexMapping](https://docs.mfem.org/html/classmfem_1_1Mesh.html#a132ae1602643c709b3e79565af061614) method. Presumably this does a topological vertex mapping, so some boundary elements become interior faces.

When particles hit a periodic boundary, we need to update their coordinates to wrap them around to the other side of the domain. At the moment, we get the boundary elements by simply looping over the `GetNBE` boundary elements. However, if "periodic-izing" an existing mesh deletes some boundary elements and converts them to interior faces, the existing method may skip over those, even though we still need to do something to particles that hit them. Is there an easy way to recover the boundary elements that have been turned into interior faces to do the periodic mapping?

Worst-case scenario, I'm thinking we could store the centroids of all the boundary elements before doing the periodic mapping, then after the mapping is done, comparing all the interior face centroids to the list of formerly boundary element centroids. But this is maybe less than robust.

**Comment by pazner** [comment]
Hello @logantm2, when a periodic mesh is created (using `Mesh::MakePeriodic` after `Mesh::CreatePeriodicVertexMapping`) the (formerly) boundary elements are not deleted, and they remain in the mesh unless you explicitly call `Mesh::RemoveInternalBoundaries`.

**Comment by vladotomov** [comment]
Can't you use the coordinates of the face-neighbor element once the particle is about to move over?

**Comment by logantm2** [comment]
> Hello @logantm2, when a periodic mesh is created (using `Mesh::MakePeriodic` after `Mesh::CreatePeriodicVertexMapping`) the (formerly) boundary elements are not deleted, and they remain in the mesh unless you explicitly call `Mesh::RemoveInternalBoundaries`.

Thanks, this is helpful, I didn't realize that they stuck around. Does this mean that two boundary elements map to the same face index under `Mesh::GetBdrElementEdgeIndex`? Or does that one face have two valid face indices?

**Comment by logantm2** [comment]
> Can't you use the coordinates of the face-neighbor element once the particle is about to move over?

With the way we handle particles hitting a boundary, the particle is held exactly on the face while the boundary effect is evaluated. In this case, the effect would be to map coordinates from a face to coordinates on another face (or I guess, the same face, but wrapped around the domain). So we could do something similar: find the centroids of the face from either side, take the difference to get a coordinate "shift" vector, then add that shift to particle coordinates when they hit the face (or subtract, depending on which side they hit from).

**Comment by pazner** [comment]
> Thanks, this is helpful, I didn't realize that they stuck around. Does this mean that two boundary elements map to the same face index under `Mesh::GetBdrElementEdgeIndex`? Or does that one face have two valid face indices?

I believe that the two boundary elements will map to the same face index, but it would be worth checking with a simple example, e.g. with a simple 2D Cartesian grid (see [here](https://mfem.org/howto/periodic-boundaries/#creating-a-periodic-mesh-by-identifying-vertices)).
```c++
// Create a 10x10 quad mesh of the unit square;
Mesh mesh = Mesh::MakeCartesian2D(10, 10, Element::QUADRILATERAL);
// Create translation vectors defining the periodicity
Vector x_translation({1.0, 0.0});
Vector y_translation({0.0, 1.0});
std::vector<Vector> translations = {x_translation, y_translation};
// Create the periodic mesh using the vertex mapping defined by the translation vectors
Mesh periodic_mesh = Mesh::MakePeriodic(mesh, mesh.CreatePeriodicVertexMapping(translations));
```

**Comment by logantm2** [comment]
Gotcha, yep, it looks like two boundary elements map to the same face index. 

This strategy seems to work well for serial meshes, but I'm running into some issues with parallel meshes.
When running the following code (which just makes a simple 2D 3x3 quad mesh that's periodic in the x-direction)
```c++
int rank;
MPI_Comm_rank(MPI_COMM_WORLD, &rank);
mfem::Mesh mfem_mesh = mfem::Mesh::MakeCartesian2D(3, 3, mfem::Element::QUADRILATERAL, true);
mfem::Vector x1_translation({1.0, 0.0});
std::vector<mfem::Vector> translations = {x1_translation};
mfem::Mesh periodic_mesh = mfem::Mesh::MakePeriodic(mfem_mesh, mfem_mesh.CreatePeriodicVertexMapping(translations));
mfem::ParMesh par_mesh(MPI_COMM_WORLD, periodic_mesh);

for (int i=0; i<par_mesh.GetNBE(); i++) {
    int face_no = par_mesh.GetBdrElementEdgeIndex(i);
    mfem::Vector centroid(2);
    mfem::ElementTransformation *trans = par_mesh.GetFaceTransformation(face_no);
    trans->SetIntPoint(&mfem::Geometries.GetCenter(trans->GetGeometryType()));
    trans->Transform(trans->GetIntPoint(), centroid);
    std::cout << "rank " << rank << " says face " << face_no << " with coords " << centroid[0] << " " << centroid[1] << " is on the boundary" << std::endl;
}
```
with 2 MPI ranks,
the mesh is decomposed as
```
* --- * --- * --- *
|  1  |  1  |  0  |
* --- * --- * --- *
|  1  |  0  |  0  |
* --- * --- * --- *
|  1  |  0  |  1  |
* --- * --- * --- *
```
so I'd expect rank 0 to have 4 boundary elements. However, the output from rank 0 is
```
rank 0 says face 10 with coords 0.5 0 is on the boundary
rank 0 says face 2 with coords 0.833333 1 is on the boundary
```
so it looks like rank 0 is not seeing the boundary elements that are shared. Rank 1 is correctly seeing all of its boundary elements. Presumably this is because, when the mesh is distributed across the communicator, each boundary element is distributed to exactly one MPI proc, so shared interior boundary elements are only seen by one proc.

**Comment by pazner** [comment]
Yes, I think you're right, `ParMesh` will only create boundary elements that correspond to faces that each partition "owns", so some of the duplicated boundary elements will disappear during the partitioning. You might need to do some of the book-keeping yourself using the shared face (/face neighbor) information.

**Comment by vladotomov** [comment]
> > Can't you use the coordinates of the face-neighbor element once the particle is about to move over?
> 
> With the way we handle particles hitting a boundary, the particle is held exactly on the face while the boundary effect is evaluated. In this case, the effect would be to map coordinates from a face to coordinates on another face (or I guess, the same face, but wrapped around the domain). So we could do something similar: find the centroids of the face from either side, take the difference to get a coordinate "shift" vector, then add that shift to particle coordinates when they hit the face (or subtract, depending on which side they hit from).

For each face, you have:
- the coordinates of the nodes, from the current element (from the L2 `Nodes` of the mesh).
- the face neighbor, even in parallel.
- the coordinates of the face neighbor element (from the L2 `Nodes` of the mesh).

Aren't those sufficient?

**Comment by logantm2** [comment]
> > > Can't you use the coordinates of the face-neighbor element once the particle is about to move over?
> > 
> > 
> > With the way we handle particles hitting a boundary, the particle is held exactly on the face while the boundary effect is evaluated. In this case, the effect would be to map coordinates from a face to coordinates on another face (or I guess, the same face, but wrapped around the domain). So we could do something similar: find the centroids of the face from either side, take the difference to get a coordinate "shift" vector, then add that shift to particle coordinates when they hit the face (or subtract, depending on which side they hit from).
> 
> For each face, you have:
> 
> * the coordinates of the nodes, from the current element (from the L2 `Nodes` of the mesh).
> * the face neighbor, even in parallel.
> * the coordinates of the face neighbor element (from the L2 `Nodes` of the mesh).
> 
> Aren't those sufficient?

So far, I've only been able to get the coordinates of one "version" of a periodic face -- that is, after two faces are merged into one via a periodic vertex mapping, I'm seemingly only able to get the coordinates of the merged face.

I can get the elements on both side of a periodic face with no problems, but I really need a face-to-face mapping, not an element-to-element one.

**Comment by logantm2** [comment]
> Yes, I think you're right, `ParMesh` will only create boundary elements that correspond to faces that each partition "owns", so some of the duplicated boundary elements will disappear during the partitioning. You might need to do some of the book-keeping yourself using the shared face (/face neighbor) information.

I've been playing around with using the DG trick from [this issue](https://github.com/mfem/mfem/issues/2768) to grab some info from face neighbors, but haven't yet figured out exactly what info I should be putting in. 

At first I tried adding a simple bool, true if the face neighbor abuts a boundary element (which should mean that it's periodic) and false otherwise. But this doesn't work if a single element shares periodic *and* non-periodic faces. This situation actually happens in the above example -- the middle-left element, owned by rank 1, shares both a periodic and a non-periodic face with rank 0.

WCS I think I can just record the centroids of each periodic face before partitioning the mesh. Then after partitioning, I can check each shared face against the recorded centroids to see which ones are periodic.

**Comment by logantm2** [comment]
For the time being, we're really only concerned with meshes that are periodic in the three cartesian directions. So for now, it looks like I can get away with saying that if a shared face's centroid lies on the surface of the mesh bounding box, it's a periodic face. And also interior boundary elements are periodic faces.

This works well enough. I may come back to thinking about this if we want anything more sophisticated.
